File: scenarios/service-bus-typescript-send-receive-messages/golden/src/service-bus.ts
import { DefaultAzureCredential } from "@azure/identity";
import {
  ServiceBusClient,
  type ProcessErrorArgs,
  type ServiceBusMessage,
  type ServiceBusReceivedMessage,
} from "@azure/service-bus";

declare const process: {
  env: Record<string, string | undefined>;
};

function requireEnvironment(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Set ${name} before running the application.`);
  }
  return value;
}

function wait(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

async function main(): Promise<void> {
  const namespace = requireEnvironment(
    "SERVICE_BUS_FULLY_QUALIFIED_NAMESPACE",
  );
  const queueName = requireEnvironment("SERVICE_BUS_QUEUE_NAME");
  const topicName = requireEnvironment("SERVICE_BUS_TOPIC_NAME");
  const subscriptionName = requireEnvironment(
    "SERVICE_BUS_SUBSCRIPTION_NAME",
  );
  const client = new ServiceBusClient(
    namespace,
    new DefaultAzureCredential(),
  );
  let queueSender:
    | ReturnType<ServiceBusClient["createSender"]>
    | undefined;
  let queueReceiver:
    | ReturnType<ServiceBusClient["createReceiver"]>
    | undefined;
  let processorReceiver:
    | ReturnType<ServiceBusClient["createReceiver"]>
    | undefined;
  let topicSender:
    | ReturnType<ServiceBusClient["createSender"]>
    | undefined;
  let subscriptionReceiver:
    | ReturnType<ServiceBusClient["createReceiver"]>
    | undefined;
  let processorSubscription:
    | { close(): Promise<void> }
    | undefined;

  try {
    queueSender = client.createSender(queueName);
    queueReceiver = client.createReceiver(queueName);
    processorReceiver = client.createReceiver(queueName);
    topicSender = client.createSender(topicName);
    subscriptionReceiver = client.createReceiver(
      topicName,
      subscriptionName,
    );
    await queueSender.sendMessages({
      body: "single queue message",
    });

    const batch = await queueSender.createMessageBatch();
    for (let index = 0; index < 5; index += 1) {
      const message: ServiceBusMessage = {
        body: `queue batch message ${index}`,
      };
      if (!batch.tryAddMessage(message)) {
        throw new Error(`Queue batch message ${index} did not fit.`);
      }
    }
    await queueSender.sendMessages(batch);

    const queueMessages = await queueReceiver.receiveMessages(5, {
      maxWaitTimeInMs: 5_000,
    });
    for (const message of queueMessages) {
      console.log(message.body);
      await queueReceiver.completeMessage(message);
    }

    const processMessage = async (
      message: ServiceBusReceivedMessage,
    ): Promise<void> => {
      if (!processorReceiver) {
        throw new Error("The processor receiver is not initialized.");
      }
      console.log(message.body);
      await processorReceiver.completeMessage(message);
    };
    const processError = async (args: ProcessErrorArgs): Promise<void> => {
      console.error(args.error);
    };
    const processorOptions = {
      autoCompleteMessages: false,
    } as const;
    processorSubscription = processorReceiver.subscribe({
      processMessage,
      processError,
    }, processorOptions);
    await wait(5_000);

    const topicMessage: ServiceBusMessage = {
      body: "topic message",
    };
    await topicSender.sendMessages(topicMessage);
    const subscriptionMessages =
      await subscriptionReceiver.receiveMessages(1, {
        maxWaitTimeInMs: 5_000,
      });
    for (const message of subscriptionMessages) {
      console.log(message.body);
      await subscriptionReceiver.completeMessage(message);
    }
  } finally {
    await Promise.allSettled([
      processorSubscription?.close(),
    ]);
    await Promise.allSettled([
      queueSender?.close(),
      queueReceiver?.close(),
      processorReceiver?.close(),
      topicSender?.close(),
      subscriptionReceiver?.close(),
    ]);
    await client.close();
  }
}

await main();

File: scenarios/service-bus-python-send-receive-messages/golden/service_bus.py
import os

from azure.servicebus import ServiceBusClient, ServiceBusMessage
from azure.servicebus.exceptions import MessageSizeExceededError


def main() -> None:
    connection_string = os.environ["SERVICE_BUS_CONNECTION_STRING"]
    queue_name = os.environ["SERVICE_BUS_QUEUE_NAME"]
    topic_name = os.environ["SERVICE_BUS_TOPIC_NAME"]
    subscription_name = os.environ["SERVICE_BUS_SUBSCRIPTION_NAME"]

    with ServiceBusClient.from_connection_string(connection_string) as client:
        with client.get_queue_sender(queue_name=queue_name) as sender:
            sender.send_messages(ServiceBusMessage("standalone queue message"))

            batch = sender.create_message_batch()
            for index in range(5):
                try:
                    batch.add_message(ServiceBusMessage(f"batch message {index}"))
                except MessageSizeExceededError:
                    raise RuntimeError("A batch message did not fit") from None
            sender.send_messages(batch)

        with client.get_queue_receiver(
            queue_name=queue_name,
            max_wait_time=5,
        ) as receiver:
            messages = receiver.receive_messages(
                max_message_count=5,
                max_wait_time=5,
            )
            for message in messages:
                print(message.body)
                receiver.complete_message(message)

        with client.get_topic_sender(topic_name=topic_name) as sender:
            sender.send_messages(ServiceBusMessage("topic message"))

        with client.get_subscription_receiver(
            topic_name=topic_name,
            subscription_name=subscription_name,
            max_wait_time=5,
        ) as receiver:
            messages = receiver.receive_messages(
                max_message_count=5,
                max_wait_time=5,
            )
            for message in messages:
                print(message.body)
                receiver.complete_message(message)


if __name__ == "__main__":
    main()
